File: lib/py/japi/internal_parse.py
import json
from typing import Dict
from typing import List, Dict, Union
import re
from typing import List, Dict, Any, Union

from japi.internal_types import Definition, Enum, ErrorDefinition, FieldDeclaration, FieldNameAndFieldDeclaration, FunctionDefinition, Japi, JsonAny, JsonArray, JsonBoolean, JsonInteger, JsonNull, JsonNumber, JsonObject, JsonString, Struct, TitleDefinition, TypeDeclaration, TypeDefinition
from japi.japi_parse_error import JapiParseError


def new_japi(japi_as_json: str) -> Japi:
    parsed_definitions: Dict[str, Definition] = {}

    japi_as_json_python: Dict[str, Any] = json.loads(japi_as_json)

    for definition_key, definition_value in japi_as_json_python.items():
        if definition_key not in parsed_definitions:
            definition = parse_definition(
                japi_as_json_python, parsed_definitions, definition_key)
            parsed_definitions[definition_key] = definition

    return Japi(japi_as_json_python, parsed_definitions)


def parse_definition(
    japi_as_json_java: Dict[str, List[object]],
    parsed_definitions: Dict[str, Definition],
    definition_key: str
) -> Definition:
    definition_as_json_java_with_doc = japi_as_json_java.get(definition_key)
    if definition_as_json_java_with_doc is None:
        raise JapiParseError(f"Could not find definition for {definition_key}")

    regex = re.compile(
        "^(struct|union|enum|error|function|event|info).([a-zA-Z_]+[a-zA-Z0-9_]*)$")
    matcher = regex.match(definition_key)
    keyword = matcher.group(1)
    definition_name = matcher.group(2)

    definition_as_json_java = definition_as_json_java_with_doc[1]

    if keyword == "function":
        try:
            definition_array = definition_as_json_java
            input_definition_as_json_java = definition_array[0]
        except (IndexError, TypeError):
            raise JapiParseError(
                f"Invalid function definition for {definition_key}")

        input_fields: Dict[str, FieldDeclaration] = {}
        for field_declaration, type_declaration_value in input_definition_as_json_java.items():
            parsed_field = parse_field(
                japi_as_json_java, parsed_definitions, field_declaration, type_declaration_value, False)
            input_fields[parsed_field.field_name] = parsed_field.field_declaration

        try:
            output_definition_as_json_java = definition_array[2]
        except (IndexError, TypeError):
            raise JapiParseError(
                f"Invalid function definition for {definition_key}")

        output_fields: Dict[str, FieldDeclaration] = {}
        for field_declaration, type_declaration_value in output_definition_as_json_java.items():
            parsed_field = parse_field(
                japi_as_json_java, parsed_definitions, field_declaration, type_declaration_value, False)
            output_fields[parsed_field.field_name] = parsed_field.field_declaration

        errors: List[str] = []
        if len(definition_array) == 4:
            error_definition = definition_array[3]

            try:
                errors = list(error_definition)
            except TypeError:
                raise JapiParseError(
                    f"Invalid function definition for {definition_key}")

            for error_def in errors:
                if error_def not in japi_as_json_java:
                    raise JapiParseError(
                        f"Unknown error reference for {error_def}")

        input_struct = Struct(f"{definition_key}.input", input_fields)
        output_struct = Struct(f"{definition_key}.output", output_fields)

        return FunctionDefinition(definition_key, input_struct, output_struct, errors)

    elif keyword == "struct":
        try:
            struct_definition_as_json_java = definition_as_json_java
        except TypeError:
            raise JapiParseError(
                f"Invalid struct definition for {definition_key}")

        fields: Dict[str, FieldDeclaration] = {}
        for field_declaration, type_declaration_value in struct_definition_as_json_java.items():
            parsed_field = parse_field(
                japi_as_json_java, parsed_definitions, field_declaration, type_declaration_value, False)
            fields[parsed_field.field_name] = parsed_field.field_declaration

        type_ = Struct(definition_key, fields)

        return TypeDefinition(definition_key, type_)

    elif keyword == "error":
        try:
            error_definition_as_json_java = definition_as_json_java
        except TypeError:
            raise JapiParseError(
                f"Invalid error definition for {definition_key}")

        fields: Dict[str, FieldDeclaration] = {}
        for field_declaration, type_declaration_value in error_definition_as_json_java.items():
            parsed_field = parse_field(
                japi_as_json_java, parsed_definitions, field_declaration, type_declaration_value, False)
            fields[parsed_field.field_name] = parsed_field.field_declaration

        return ErrorDefinition(definition_key, fields)

    elif keyword == "enum":
        try:
            enum_definition_as_json_java = definition_as_json_java
        except TypeError:
            raise JapiParseError(
                f"Invalid enum definition for {definition_key}")

        cases: Dict[str, Struct] = {}
        for enum_case, case_struct_definition_as_java in enum_definition_as_json_java.items():
            try:
                case_struct_definition_as_java = dict(
                    case_struct_definition_as_java)
            except TypeError:
                raise JapiParseError(
                    f"Invalid enum definition for {definition_key}")

            fields: Dict[str, FieldDeclaration] = {}
            for case_struct_field_declaration, case_struct_type_declaration_value in case_struct_definition_as_java.items():
                case_struct_parsed_field = parse_field(
                    japi_as_json_java, parsed_definitions, case_struct_field_declaration, case_struct_type_declaration_value, False)
                fields[case_struct_parsed_field.field_name] = case_struct_parsed_field.field_declaration

            struct = Struct(f"{definition_key}.{enum_case}", fields)
            cases[enum_case] = struct

        type_ = Enum(definition_key, cases)

        return TypeDefinition(definition_key, type_)

    elif keyword == "info":
        return TitleDefinition(definition_key)

    else:
        raise JapiParseError(f"Unrecognized japi keyword {keyword}")


def parse_field(japi_as_json_java: Dict[str, List[Any]], parsed_definitions: Dict[str, Any], field_declaration: str, type_declaration_value: Any, is_for_union: bool) -> FieldNameAndFieldDeclaration:
    regex = re.compile(r"^([a-zA-Z_]+[a-zA-Z0-9_]*)(!)?$")
    matcher = regex.match(field_declaration)

    if not matcher:
        raise JapiParseError("Invalid field declaration: %s" %
                             field_declaration)

    field_name = matcher.group(1)
    optional = bool(matcher.group(2))

    if optional and is_for_union:
        raise JapiParseError("Union keys cannot be marked as optional")

    try:
        type_declaration_string = str(type_declaration_value)
    except TypeError:
        raise JapiParseError("Type declarations should be strings")

    type_declaration = parse_type(
        japi_as_json_java, parsed_definitions, type_declaration_string)

    return FieldNameAndFieldDeclaration(field_name, FieldDeclaration(type_declaration, optional))


def parse_type(japi_as_json_java: Dict[str, List[Any]], parsed_definitions: Dict[str, Any], type_declaration: str) -> TypeDeclaration:
    regex = re.compile(
        r"^((null|boolean|integer|number|string|any)|((array|object)(<(.*)>)?)|((enum|struct)\.([a-zA-Z_]\w*)))(\?)?$")
    matcher = regex.match(type_declaration)

    if not matcher:
        raise JapiParseError("Invalid type declaration: %s" % type_declaration)

    nullable = bool(matcher.group(10))

    try:
        name = matcher.group(2)
        if name is None:
            raise RuntimeError("Ignore: will try another type")

        if name == "null":
            if nullable:
                raise JapiParseError("Cannot declare null type as nullable")
            nullable = True
            return TypeDeclaration(JsonNull(), nullable)

        type_obj = None
        if name == "boolean":
            type_obj = JsonBoolean()
        elif name == "integer":
            type_obj = JsonInteger()
        elif name == "number":
            type_obj = JsonNumber()
        elif name == "string":
            type_obj = JsonString()
        elif name == "any":
            type_obj = JsonAny()
        else:
            raise JapiParseError("Unrecognized type: %s" % name)

        return TypeDeclaration(type_obj, nullable)
    except JapiParseError as e1:
        raise e1
    except Exception:
        pass

    try:
        name = matcher.group(4)
        if name is None:
            raise RuntimeError("Ignore: will try another type")

        nested_name = matcher.group(6)
        nested_type = None
        if nested_name:
            nested_type = parse_type(
                japi_as_json_java, parsed_definitions, nested_name)
        else:
            nested_type = TypeDeclaration(JsonAny(), False)

        type_obj = None
        if name == "array":
            type_obj = JsonArray(nested_type)
        elif name == "object":
            type_obj = JsonObject(nested_type)
        else:
            raise JapiParseError("Unrecognized type: %s" % name)

        return TypeDeclaration(type_obj, nullable)
    except JapiParseError as e1:
        raise e1
    except Exception:
        pass

    name = matcher.group(7)
    if name is None:
        raise RuntimeError("Ignore: will try another type")

    definition = parsed_definitions.setdefault(name, parse_definition(
        japi_as_json_java, parsed_definitions, name))
    if isinstance(definition, TypeDefinition):
        return TypeDeclaration(definition.type, nullable)
    elif isinstance(definition, FunctionDefinition):
        raise JapiParseError(
            "Cannot reference a function in type declarations")
    elif isinstance(definition, ErrorDefinition):
        raise JapiParseError(
            "Cannot reference an error in type declarations")
    else:
        raise JapiParseError("Unknown definition: %s" % type_declaration)

File: lib/java/src/main/java/io/github/brenbar/japi/InternalParse.java
package io.github.brenbar.japi;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.TreeSet;
import java.util.regex.Pattern;

class InternalParse {

    static JApiSchemaTuple newJApiSchemaWithInternalSchema(String jApiSchemaAsJson,
            Map<String, TypeExtension> typeExtensions) {
        var combinedSchemaJson = combineJsonSchemas(List.of(
                jApiSchemaAsJson,
                InternalJApi.getJson()));
        var schema = newJApiSchema(combinedSchemaJson, typeExtensions);

        return schema;
    }

    static String combineJsonSchemas(List<String> jsonSchemas) {
        var objectMapper = new ObjectMapper();
        var finalJsonList = new ArrayList<Object>();

        var pseudoJsonSchemas = new ArrayList<Object>();
        for (var jsonSchema : jsonSchemas) {
            List<Object> pseudoJsonSchema;
            try {
                pseudoJsonSchema = objectMapper.readValue(jsonSchema, new TypeReference<List<Object>>() {
                });
            } catch (JsonProcessingException e) {
                throw new JApiSchemaParseError("Document root must be an array of objects", e);
            }
            pseudoJsonSchemas.add(pseudoJsonSchema);
        }

        var jsonSchemaKeys = new HashSet<String>();
        var duplicatedJsonSchemaKeys = new HashSet<String>();
        for (var pseudoJsonSchema : pseudoJsonSchemas) {
            var definitions = (List<Object>) pseudoJsonSchema;
            for (var definition : definitions) {
                var definitionAsPsuedoJson = (Map<String, Object>) definition;
                var key = findSchemaKey(definitionAsPsuedoJson);
                if (jsonSchemaKeys.contains(key)) {
                    duplicatedJsonSchemaKeys.add(key);
                } else {
                    jsonSchemaKeys.add(key);
                }
            }
        }

        if (!duplicatedJsonSchemaKeys.isEmpty()) {
            var sortedKeys = new TreeSet<String>(duplicatedJsonSchemaKeys);
            throw new JApiSchemaParseError(
                    "Final schema has duplicate keys: %s".formatted(sortedKeys));
        }

        for (var pseudoJsonSchema : pseudoJsonSchemas) {
            var definitions = (List<Object>) pseudoJsonSchema;
            finalJsonList.addAll(definitions);
        }

        try {
            return objectMapper.writeValueAsString(finalJsonList);
        } catch (JsonProcessingException e) {
            throw new JApiSchemaParseError("Could not combine schemas", e);
        }
    }

    private static String findSchemaKey(Map<String, Object> definition) {
        for (var e : definition.keySet()) {
            if (e.matches("^(struct|enum|fn|trait|info|ext)\\..*")) {
                return e;
            }
        }
        throw new JApiSchemaParseError(
                "Invalid definition. Each definition should have one key matching the regex ^(struct|enum|fn|trait|info|ext)\\..* but was %s"
                        .formatted(definition));
    }

    static JApiSchemaTuple newJApiSchema(String jApiSchemaAsJson, Map<String, TypeExtension> typeExtensions) {
        var parsedTypes = new HashMap<String, Type>();

        var objectMapper = new ObjectMapper();
        List<Map<String, Object>> initialJapiSchemaAsParsedJson;
        try {
            initialJapiSchemaAsParsedJson = objectMapper.readValue(jApiSchemaAsJson, new TypeReference<>() {
            });
        } catch (IOException e) {
            throw new JApiSchemaParseError("Document root must be an array of objects", e);
        }

        var jApiSchemaAsParsedJson = new HashMap<String, Object>();

        for (var definition : initialJapiSchemaAsParsedJson) {
            String schemaKey = findSchemaKey(definition);
            jApiSchemaAsParsedJson.put(schemaKey, definition);
        }

        var schemaKeys = jApiSchemaAsParsedJson.keySet();

        var traitSchemaKeys = new HashSet<String>();

        for (var schemaKey : schemaKeys) {
            if (schemaKey.startsWith("trait.")) {
                traitSchemaKeys.add(schemaKey);
                continue;
            }

            getOrParseType(schemaKey, jApiSchemaAsParsedJson, parsedTypes, typeExtensions);
        }

        // Apply trait to all functions
        for (var traitSchemaKey : traitSchemaKeys) {

            var traitDefinition = (Map<String, Object>) jApiSchemaAsParsedJson.get(traitSchemaKey);

            var trait = parseTraitType(traitDefinition, traitSchemaKey, jApiSchemaAsParsedJson,
                    parsedTypes,
                    typeExtensions);

            applyTraitToParsedTypes(trait, parsedTypes);

            parsedTypes.put(trait.name, trait);
        }

        // Ensure all type extensions are defined
        for (var entry : typeExtensions.entrySet()) {
            var typeExtensionName = entry.getKey();
            var typeExtension = (Ext) parsedTypes.get(typeExtensionName);
            if (typeExtension == null) {
                throw new JApiSchemaParseError("Undefined extension %s".formatted(typeExtensionName));
            }
        }

        return new JApiSchemaTuple((Map<String, Object>) (Object) jApiSchemaAsParsedJson, parsedTypes);
    }

    static void applyTraitToParsedTypes(Trait trait, Map<String, Type> parsedTypes) {
        for (var parsedType : parsedTypes.entrySet()) {
            Fn f;
            try {
                f = (Fn) parsedType.getValue();
            } catch (ClassCastException e) {
                continue;
            }

            var regex = Pattern.compile(trait.regex);
            var matcher = regex.matcher(f.name);
            if (!matcher.find()) {
                continue;
            }

            if (f.name.startsWith("fn._")) {
                // Only internal traits can change internal functions
                if (!trait.name.startsWith("trait._")) {
                    continue;
                }
            }

            for (var traitArgumentField : trait.fn.arg.fields.entrySet()) {
                var newKey = traitArgumentField.getKey();
                if (f.arg.fields.containsKey(newKey)) {
                    throw new JApiSchemaParseError(
                            "Argument field already in use: %s".formatted(newKey));
                }
                f.arg.fields.put(newKey, traitArgumentField.getValue());
            }

            for (var traitResultField : trait.fn.result.values.entrySet()) {
                var newKey = traitResultField.getKey();
                if (f.result.values.containsKey(newKey)) {
                    throw new JApiSchemaParseError(
                            "Result value already in use: %s".formatted(newKey));
                }
                f.result.values.put(newKey, traitResultField.getValue());
            }
        }
    }

    public static Trait parseTraitType(
            Map<String, Object> traitDefinitionAsParsedJson,
            String definitionKey,
            Map<String, Object> jApiSchemaAsParsedJson,
            Map<String, Type> parsedTypes,
            Map<String, TypeExtension> typeExtensions) {
        Map<String, Object> def;
        try {
            def = (Map<String, Object>) traitDefinitionAsParsedJson.get(definitionKey);
        } catch (ClassCastException e) {
            throw new JApiSchemaParseError("Invalid trait definition %s".formatted(definitionKey));
        }

        String traitFunctionRegex;
        String traitFunctionKey;
        if (def.containsKey("fn.*")) {
            traitFunctionKey = "fn.*";
            traitFunctionRegex = "^fn\\.[a-zA-Z]";
        } else if (def.containsKey("fn._?*")) {
            if (!definitionKey.startsWith("trait._")) {
                throw new JApiSchemaParseError("Invalid trait definition %s".formatted(definitionKey));
            }
            traitFunctionKey = "fn._?*";
            traitFunctionRegex = "^fn\\.[a-zA-Z_]";
        } else {
            throw new JApiSchemaParseError("Invalid trait definition %s".formatted(definitionKey));
        }

        var traitFunction = parseFunctionType(def, traitFunctionKey, jApiSchemaAsParsedJson, parsedTypes,
                typeExtensions,
                true);

        return new Trait(definitionKey, traitFunction, traitFunctionRegex);
    }

    private static Fn parseFunctionType(
            Map<String, Object> functionDefinitionAsParsedJson,
            String definitionKey,
            Map<String, Object> jApiSchemaAsParsedJson,
            Map<String, Type> parsedTypes,
            Map<String, TypeExtension> typeExtensions,
            boolean isForTrait) {
        Map<String, Object> argumentDefinitionAsParsedJson;
        try {
            argumentDefinitionAsParsedJson = (Map<String, Object>) functionDefinitionAsParsedJson.get(definitionKey);
        } catch (ClassCastException e) {
            throw new JApiSchemaParseError("Invalid function definition for %s".formatted(definitionKey));
        }
        var argumentFields = new HashMap<String, FieldDeclaration>();
        for (var entry : argumentDefinitionAsParsedJson.entrySet()) {
            var fieldDeclaration = entry.getKey();
            var typeDeclarationValue = entry.getValue();
            var parsedField = parseField(fieldDeclaration,
                    typeDeclarationValue, false, jApiSchemaAsParsedJson, parsedTypes, typeExtensions);
            argumentFields.put(parsedField.fieldName, parsedField.fieldDeclaration);
        }

        var argType = new Struct(definitionKey, argumentFields);

        Map<String, Object> resultDefinitionAsParsedJson;
        try {
            resultDefinitionAsParsedJson = (Map<String, Object>) functionDefinitionAsParsedJson.get("->");
        } catch (ClassCastException e) {
            throw new JApiSchemaParseError("Invalid function definition for %s".formatted(definitionKey));
        }

        if (!isForTrait) {
            if (!resultDefinitionAsParsedJson.containsKey("ok")) {
                throw new JApiSchemaParseError("Invalid function definition for %s".formatted(definitionKey));
            }
        }

        var values = new HashMap<String, Struct>();
        for (var entry : resultDefinitionAsParsedJson.entrySet()) {
            Map<String, Object> enumValueData;
            try {
                enumValueData = (Map<String, Object>) entry.getValue();
            } catch (ClassCastException e) {
                throw new JApiSchemaParseError("Invalid function definition for %s".formatted(definitionKey));
            }
            var enumValue = entry.getKey();

            var fields = new HashMap<String, FieldDeclaration>();
            for (var structEntry : enumValueData.entrySet()) {
                var fieldDeclaration = structEntry.getKey();
                var typeDeclarationValue = structEntry.getValue();
                var parsedField = parseField(fieldDeclaration,
                        typeDeclarationValue, false, jApiSchemaAsParsedJson, parsedTypes, typeExtensions);
                fields.put(parsedField.fieldName, parsedField.fieldDeclaration);
            }

            var enumStruct = new Struct("->.%s".formatted(enumValue), fields);

            values.put(enumValue, enumStruct);
        }

        var resultType = new Enum("%s.->".formatted(definitionKey), values);

        var type = new Fn(definitionKey, argType, resultType);

        return type;
    }

    private static Struct parseStructType(
            Map<String, Object> structDefinitionAsParsedJson,
            String definitionKey,
            Map<String, Object> jApiSchemaAsParsedJson,
            Map<String, Type> parsedTypes,
            Map<String, TypeExtension> typeExtensions) {
        var definition = (Map<String, Object>) structDefinitionAsParsedJson.get(definitionKey);

        var fields = new HashMap<String, FieldDeclaration>();
        for (var entry : definition.entrySet()) {
            var fieldDeclaration = entry.getKey();
            var typeDeclarationValue = entry.getValue();
            var parsedField = parseField(fieldDeclaration,
                    typeDeclarationValue, false, jApiSchemaAsParsedJson, parsedTypes, typeExtensions);
            fields.put(parsedField.fieldName, parsedField.fieldDeclaration);
        }

        var type = new Struct(definitionKey, fields);

        return type;
    }

    private static Enum parseEnumType(
            Map<String, Object> enumDefinitionAsParsedJson,
            String definitionKey,
            Map<String, Object> jApiSchemaAsParsedJson,
            Map<String, Type> parsedTypes,
            Map<String, TypeExtension> typeExtensions) {
        var definition = (Map<String, Object>) enumDefinitionAsParsedJson.get(definitionKey);

        var values = new HashMap<String, Struct>();
        for (var entry : definition.entrySet()) {
            Map<String, Object> enumStructData;
            try {
                enumStructData = (Map<String, Object>) entry.getValue();
            } catch (ClassCastException e) {
                throw new JApiSchemaParseError("Invalid enum definition for %s".formatted(definitionKey));
            }
            var enumValue = entry.getKey();

            var fields = new HashMap<String, FieldDeclaration>();
            for (var structEntry : enumStructData.entrySet()) {
                var fieldDeclaration = structEntry.getKey();
                var typeDeclarationValue = structEntry.getValue();
                var parsedField = parseField(fieldDeclaration,
                        typeDeclarationValue, false, jApiSchemaAsParsedJson, parsedTypes, typeExtensions);
                fields.put(parsedField.fieldName, parsedField.fieldDeclaration);
            }

            var enumStruct = new Struct("%s.%s".formatted(definitionKey, enumValue), fields);

            values.put(enumValue, enumStruct);
        }

        var type = new Enum(definitionKey, values);

        return type;
    }

    private static FieldNameAndFieldDeclaration parseField(
            String fieldDeclaration,
            Object typeDeclarationValue,
            boolean isForEnum,
            Map<String, Object> jApiSchemaAsParsedJson,
            Map<String, Type> parsedTypes,
            Map<String, TypeExtension> typeExtensions) {
        var regex = Pattern.compile("^([a-zA-Z_]+[a-zA-Z0-9_]*)(!)?$");
        var matcher = regex.matcher(fieldDeclaration);
        if (!matcher.find()) {
            throw new JApiSchemaParseError("Could not parse field declaration: %s".formatted(fieldDeclaration));
        }

        String fieldName = matcher.group(1);

        boolean optional = matcher.group(2) != null;

        String typeDeclarationString;
        try {
            typeDeclarationString = (String) typeDeclarationValue;
        } catch (ClassCastException e) {
            throw new JApiSchemaParseError(
                    "Type declarations should be strings: %s %s".formatted(fieldDeclaration, typeDeclarationValue));
        }

        if (optional && isForEnum) {
            throw new JApiSchemaParseError("Enum keys cannot be marked as optional");
        }

        var typeDeclaration = parseTypeDeclaration(typeDeclarationString, jApiSchemaAsParsedJson, parsedTypes,
                typeExtensions);

        return new FieldNameAndFieldDeclaration(fieldName, new FieldDeclaration(typeDeclaration, optional));
    }

    private static TypeDeclaration parseTypeDeclaration(String typeDeclaration,
            Map<String, Object> jApiSchemaAsParsedJson,
            Map<String, Type> parsedTypes, Map<String, TypeExtension> typeExtensions) {
        var regex = Pattern.compile("^(.*?)(\\?)?$");
        var matcher = regex.matcher(typeDeclaration);
        if (!matcher.find()) {
            throw new JApiSchemaParseError("Could not parse type declaration: %s".formatted(typeDeclaration));
        }

        var typeName = matcher.group(1);
        var nullable = matcher.group(2) != null;

        var type = getOrParseType(typeName, jApiSchemaAsParsedJson, parsedTypes, typeExtensions);

        return new TypeDeclaration(type, nullable);
    }

    private static Type getOrParseType(String typeName, Map<String, Object> jApiSchemaAsParsedJson,
            Map<String, Type> parsedTypes, Map<String, TypeExtension> typeExtensions) {
        var existingType = parsedTypes.get(typeName);
        if (existingType != null) {
            return existingType;
        }

        var regex = Pattern.compile(
                "^((boolean|integer|number|string|any)|((array|object)(<(.*)>)?)|((enum|struct|fn|info|ext)\\.([a-zA-Z_]\\w*)))$");
        var matcher = regex.matcher(typeName);
        if (!matcher.find()) {
            throw new JApiSchemaParseError("Unrecognized type: %s".formatted(typeName));
        }

        var standardTypeName = matcher.group(2);
        if (standardTypeName != null) {
            return switch (standardTypeName) {
                case "boolean" -> new JsonBoolean();
                case "integer" -> new JsonInteger();
                case "number" -> new JsonNumber();
                case "string" -> new JsonString();
                case "any" -> new JsonAny();
                default -> throw new JApiSchemaParseError("Unrecognized type: %s".formatted(standardTypeName));
            };
        }

        var collectionTypeName = matcher.group(4);
        if (collectionTypeName != null) {
            var nestedName = matcher.group(6);

            TypeDeclaration nestedType;
            if (nestedName != null) {
                nestedType = parseTypeDeclaration(nestedName, jApiSchemaAsParsedJson, parsedTypes, typeExtensions);
            } else {
                nestedType = new TypeDeclaration(new JsonAny(), false);
            }

            return switch (collectionTypeName) {
                case "array" -> new JsonArray(nestedType);
                case "object" -> new JsonObject(nestedType);
                default -> throw new JApiSchemaParseError("Unrecognized type: %s".formatted(collectionTypeName));
            };
        }

        var customTypeName = matcher.group(7);
        if (customTypeName != null) {
            var typePrefix = matcher.group(8);
            var definition = (Map<String, Object>) jApiSchemaAsParsedJson.get(customTypeName);
            var type = switch (typePrefix) {
                case "struct" ->
                    parseStructType(definition, customTypeName, jApiSchemaAsParsedJson, parsedTypes, typeExtensions);
                case "enum" -> parseEnumType(definition, typeName, jApiSchemaAsParsedJson, parsedTypes, typeExtensions);
                case "fn" -> parseFunctionType(definition, customTypeName, jApiSchemaAsParsedJson, parsedTypes,
                        typeExtensions, false);
                case "info" -> new Info(customTypeName);
                case "ext" -> {
                    var typeExtension = typeExtensions.get(customTypeName);
                    if (typeExtension == null) {
                        throw new JApiSchemaParseError("Type extensions must be configured in Server.Options");
                    }
                    yield new Ext(customTypeName, typeExtension);
                }
                default -> throw new JApiSchemaParseError("Unrecognized type: %s".formatted(collectionTypeName));
            };

            parsedTypes.put(customTypeName, type);

            return type;
        }

        throw new JApiSchemaParseError("Invalid type: %s".formatted(typeName));
    }
}
